scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordered_serialization/providers/TraversablesOrderedBuf.scala

```diff
@@ -0,0 +1,242 @@
+      val (tpe1, tpe2) = (outerType.asInstanceOf[TypeRefApi].args(0), outerType.asInstanceOf[TypeRefApi].args(1))
+      val containerType = typeOf[Tuple2[Any, Any]].asInstanceOf[TypeRef]
+      import compat._
+      TypeRef.apply(containerType.pre, containerType.sym, List(tpe1, tpe2))
+    } else {
+      outerType.asInstanceOf[TypeRefApi].args.head
+    }
+
+    val innerTypes = outerType.asInstanceOf[TypeRefApi].args
+
+    val innerBuf: TreeOrderedBuf[c.type] = dispatcher(innerType)
+
+    new TreeOrderedBuf[c.type] {
+      override val ctx: c.type = c
+      override val tpe = outerType
+      override def compareBinary(inputStreamA: ctx.TermName, inputStreamB: ctx.TermName) = {
+        val innerCompareFn = freshT("innerCompareFn")
+        val a = freshT("a")
+        val b = freshT("b")
+        q"""
+        def $innerCompareFn(a: _root_.java.io.InputStream, b: _root_.java.io.InputStream) = {
+          val $a = a
+          val $b = b
+          ${innerBuf.compareBinary(a, b)}
+        }
+        _root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.TraversableHelpers.rawCompare($inputStreamA, $inputStreamB)($innerCompareFn)
+      """
+      }
+
+      override def put(inputStream: ctx.TermName, element: ctx.TermName) = {
+        val bytes = freshT("bytes")
+        val len = freshT("len")
+        val innerElement = freshT("innerElement")
+        val cmpRes = freshT("cmpRes")
+        maybeSort match {
+          case DoSort =>
+            q"""
+          val $len = $element.size
+          $inputStream.writeSize($len)
+
+          if($len > 0) {
+            $element.toArray.sortWith { (a, b) =>
+                val $cmpRes = ${innerBuf.compare(newTermName("a"), newTermName("b"))}
+                $cmpRes < 0
+            }.foreach{ case $innerElement =>
+              ${innerBuf.put(inputStream, innerElement)}
+            }
+          }
+        """
+          case NoSort =>
+            q"""
+        val $len: Int = $element.size
+        $inputStream.writeSize($len)
+        $element.foreach { case $innerElement =>
+            ${innerBuf.put(inputStream, innerElement)}
+        }
+        """
+        }
+
+      }
+      override def hash(element: ctx.TermName): ctx.Tree = q"$element.hashCode"
+
+      override def get(inputStream: ctx.TermName): ctx.Tree = {
+        val len = freshT("len")
+        val firstVal = freshT("firstVal")
+        val travBuilder = freshT("travBuilder")
+        val iter = freshT("iter")
+        val extractionTree = maybeArray match {
+          case IsArray =>
+            q"""val $travBuilder = new Array[..$innerTypes]($len)
+            var $iter = 0
+            while($iter < $len) {
+              $travBuilder($iter) = ${innerBuf.get(inputStream)}
+              $iter = $iter + 1
+            }
+            $travBuilder : $outerType
+            """
+          case NotArray =>
+            q"""val $travBuilder = $companionSymbol.newBuilder[..$innerTypes]
```

Comment:
builders can take a size hint, I think. It might speed things up since we know the size.